typing/parmatch.ml

```diff
@@ -42,7 +42,158 @@ let rec omegas i =
 
 let omega_list l = List.map (fun _ -> omega) l
 
-let zero = make_pat (Tpat_constant (Const_int 0)) Ctype.none Env.empty
+module Pattern_head : sig
+  type desc =
+    | Any
+    | Construct of constructor_description
+    | Constant of constant
+    | Tuple of int
+    | Record of label_description list
+    | Variant of
+        { tag: label; has_arg: bool;
+          cstr_row: row_desc ref;
+          type_row : unit -> row_desc; }
+          (* the row of the type may evolve if [close_variant] is called,
+             hence the (unit -> ...) delay *)
```

Comment:
I find this harder to read than the previous `row_of_pat`. Could this just store `pat_env` and do `row_of_pat` as before? Is it relevant when `pat_env` is sampled?

Reply:
Our previous attempt precomputed the result of `row_of_pat`, and that didn't work at all. We could indeed store `pat_env` and `pat_type` and have a `row_of_pat` that takes both as parameters (or that takes the type of the record, hoisted out of the variant definition). Would you prefer that?

Reply:
Right. I don't understand the ordering constraints here. Are these fields mutated during parmatch?

It sounds like there's something subtle with mutation going on here, in which case the weirdness of the thunk in the record is probably a good thing: the old code was just as subtle but didn't look weird.

Reply:
> Right. I don't understand the ordering constraints here. Are these fields mutated during parmatch?

Yes, cf. `close_variant`.

Reply:
At a high level, my understanding is that if you do, say

```ocaml
function
| `A -> ()
| `B -> ()
```

then at first OCaml is going to assume when typing each clause independently that this is pattern-matching on ``[> `A | `B ]``, then at exhaustiveness-checking time (in parmatch) it is going to realize that this would be non-exhaustive (due to the absence of a catch-all clause), and "close" the variant by turning it into ``[< `A | `B ]``.

Reply:
Right, thanks, I understand now. So exhaustiveness checking has the side-effect of closing some variants, and `type_row` can return different values depending on when it's called.

In that case, I think the new code is an improvement. It doesn't look as simple as the old code, because the old code wasn't really that simple.